stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -638,14 +767,32 @@ public final class _TensorComputation {
 
   // NOTE: the following properties are intentionally not implemented as an enum
   // due to high churn, *please do not refactor for Swiftiness*.
+  private var stateTFE: TFEState?
   private var stateTF: TFState?
 
-  /// The thread to run tensor computation in. The global config flag
-  /// '_RuntimeConfig.usesSynchronousExecution' decides whether tensor
-  /// computation should be synchronous: if true, this property will always be
-  /// nil. Otherwise, this property is non-nil only when the tensor computation
-  /// is on-going.
-  private var pthread: pthread_t?
+  /// The threads to run tensor computation in. In eager mode, we use N threads
+  /// when the tensor computation involves N device functions. In non-eager
+  /// mode, we use a single thread (that's sufficient as these N device
+  /// functions can be sent to the same TF_SessionRun() call.)
+  ///
+  /// The global config flag '_RuntimeConfig.usesSynchronousExecution' decides
+  /// whether tensor computation should be synchronous: if true, this property
+  /// will always be nil. Otherwise, this property is non-nil only when the
+  /// tensor computation is on-going.
+  /// TODO: Remove the `usesSynchronousExecution`mode as it is very limiting
+  /// (e.g. does not support running multiple device functions).
+  private var pthreads: [pthread_t] = []
+
+  /// The data structure to pass into pthread creation API. 
+  class FuncHandle {
```

Comment:
I don't quite understand why a separate class is needed. Since now every `_TensorComputation` is getting a thread index (unless its sync runtime of course), can we simply add a `var threadIndex: Int?` to `_TensorComputation` to avoid an additional level of complication?

Reply:
Added a comment: We cannot have the ThreadBody closure below close over `threadIndex`, because ThreadBody is of C convention.

Reply:
I understand that this is to work around that this is a C convention function. My question was: Why not just add the index as a member in `_TensorComputation`? Then you would not need to change the original self-passing code at all.

Reply:
We are calling `_TensorComputation.execute()` in concurrent threads, so if we make `threadIndex` a property, we'll need to mutex-protect it. I think this change here is simpler.

Reply:
I see! Would it be possible to have a clearer name? I don't quite get what "Func" is referring to. Something like `_TensorComputationThreadHandle`?

Reply:
`_TensorComputationThreadHandle` is a bit too long. :-)

I settled on `ThreadParam` -- it's local to `_TensorComputation`, so the context should make its semantics clear. WDYT?

Reply:
Sure thing!

Reply:
Another minor nit: the class can be `private`.

Reply:
SG. will address in the next PR.